flesh out getting started guide (#69)
qognitive/fast-pauli: docs/getting_started.rst

Getting Started
=====================

Welcome to :code:`fast-pauli` from `Qognitive <https://www.qognitive.io/>`_, an open-source Python / C++ library for optimized operations on Pauli matrices and Pauli strings
based on `PauliComposer <https://arxiv.org/abs/2301.00560>`_.
In this guide, we'll introduce some of the important operations to help users get started as well as some conceptual background on Pauli matrices and Pauli strings.

> For more details on our programmatic interface, see the :doc:`python_api` or :doc:`cpp_api` documentation.
> 
> For tips on installing the library, check out the guide: :doc:`index`.

Reviewer: ```suggestion
If you want to follow along with this guide, please follow the installation instructions in :doc:`index`.
For more details on our programmatic interface, see the :doc:`python_api` or :doc:`cpp_api` documentation.
```

=====================

> Welcome to Fast-Pauli from `Qognitive <https://www.qognitive.io/>`_, an open-source Python / C++ library for optimized operations on Pauli matrices and Pauli strings

Reviewer: ```suggestion
Welcome to :code:`fast-pauli` from `Qognitive <https://www.qognitive.io/>`_, an open-source Python / C++ library for optimized operations on Pauli matrices and Pauli strings
```

based on `PauliComposer <https://arxiv.org/abs/2301.00560>`_.
In this guide, we'll introduce some of the important operations to help users get started as well as some conceptual background on Pauli matrices and Pauli strings.

> For more details, see the :doc:`python_api` or :doc:`cpp_api` documentation.

Reviewer: ```suggestion
For more details on our programmatic interface, see the :doc:`python_api` or :doc:`cpp_api` documentation.
```

For tips on installing the library, check out the guide: :doc:`index`.

Pauli Matrices
------------------------

For a more in-depth overview of Pauli matrices, see `here <https://en.wikipedia.org/wiki/Pauli_matrices>`_.

In math and physics, a `Pauli matrix <https://en.wikipedia.org/wiki/Pauli_matrices>`_, named after the physicist Wolfgang Pauli, is any one of the special 2 x 2 complex matrices in the set (often denoted by the greek letter :math:`\sigma`) :

.. math::

    \sigma_x = \begin{bmatrix} 0 & 1 \\ 1 & 0 \end{bmatrix}
    \sigma_y = \begin{bmatrix} 0 & -i \\ i & 0 \end{bmatrix}
    \sigma_z = \begin{bmatrix} 1 & 0 \\ 0 & -1 \end{bmatrix}

> All the pauli matrices share the properties that they are:

Reviewer: ```suggestion
All the Pauli matrices share the properties that they are:
```

=====================

Welcome to Fast-Pauli from `Qognitive <https://www.qognitive.io/>`_, an open-source Python / C++ library for optimized operations on Pauli matrices and Pauli strings
based on `PauliComposer <https://arxiv.org/abs/2301.00560>`_.
In this guide, we'll introduce some of the important operations to help users get started as well as some conceptual background on Pauli matrices and Pauli strings.

For more details, see the :doc:`python_api` or :doc:`cpp_api` documentation.

For tips on installing the library, check out the guide: :doc:`index`.

Pauli Matrices
------------------------

For a more in-depth overview of Pauli matrices, see `here <https://en.wikipedia.org/wiki/Pauli_matrices>`_.

In math and physics, a `Pauli matrix <https://en.wikipedia.org/wiki/Pauli_matrices>`_, named after the physicist Wolfgang Pauli, is any one of the special 2 x 2 complex matrices in the set (often denoted by the greek letter :math:`\sigma`) :

.. math::

>     \sigma_0 = \begin{bmatrix} 1 & 0 \\ 0 & 1 \end{bmatrix}

Reviewer 2: technically not a pauli matrix

Author: Changed to just include Sigma_X,Y,Z and then mention I=Sigma_0 after

    pauli_y = fp.Pauli('Y')
    pauli_z = fp.Pauli('Z')

    str(pauli_0)  # returns "I"

We can also multiply two ``Pauli`` objects together to get a ``Pauli`` object representing the matrix product of the two pauli matrices.

> The result includes a phase factor because the product of two Pauli matrices is not another Pauli matrix.

Reviewer 2: ```suggestion
The result includes a phase factor because the product of two Pauli matrices is not another Pauli matrix, but instead a Pauli matrix times by either `i` or `-i` which here we call the phase factor.
```

    pauli_y = fp.Pauli('Y')
    pauli_z = fp.Pauli('Z')

    str(pauli_0)  # returns "I"

> We can also multiply two ``Pauli`` objects together to get a ``Pauli`` object representing the tensor product of the two pauli matrices.

Reviewer 3: I think you meant matrix product here. you get a pair of phase and pauli object after @ing two Pauli objects

    pauli_y = fp.Pauli('Y')
    pauli_z = fp.Pauli('Z')

    str(pauli_0)  # returns "I"

We can also multiply two ``Pauli`` objects together to get a ``Pauli`` object representing the tensor product of the two pauli matrices.

> The result includes a phase factor because Pauli matrices do not necessarily commute. They exhibit the following `commutation relations <https://en.wikipedia.org/wiki/Pauli_matrices#Algebraic_properties>`_:

Reviewer 2: So it isn't the commutation that is the problem (and two pauli matrices don't commute). It is the fact that the pauli matrices are not actually closed (two pauli matrices multiplying each other do not equal another pauli matrix).

The commutation doesn't actually come up in any of the stuff here. That would just be X @ Y != Y @ X, which we don't discuss

Pauli Strings
------------------------

Pauli strings are tensor-product combinations of Pauli matrices. For example, the following is a valid Pauli string:

.. math::

    \mathcal{\hat{P}} = \sigma_x \otimes \sigma_y \otimes \sigma_z

where :math:`\otimes` denotes the tensor or `Kronecker <https://en.wikipedia.org/wiki/Kronecker_product>`_ product, and we can more simply denote by

.. math::

    \mathcal{\hat{P}} = XYZ

A Pauli string of length ``N`` is a tensor product of ``N``
Pauli matrices. A ``N``-length Pauli String is a :math:`2^N \times 2^N` matrix, so ``XYZ`` is a :math:`8 \times 8` matrix.

> In ``fast_pauli``, we represent Pauli strings using the ``PauliString`` class. For example, to construct the Pauli string ``X, Y, Z``, we can do:

Reviewer: ```suggestion
In ``fast_pauli``, we represent Pauli strings using the ``PauliString`` class, which represents this operator sparsely as just an array of `Pauli` objects. For example, to construct the Pauli string ``X\otimesY\otimesZ``, we can do:
```

Author: Since we don't bring up operators yet, could we say `represents this tensor product sparsely....`

Pauli Strings
------------------------

> Pauli strings, also known as Pauli words, are tensor-product combinations of Pauli matrices. For example, the following is a valid Pauli string:

Reviewer 2: I wouldn't mention pauli words. They'll know what it is by the math.

.. math::

    \mathcal{\hat{P}} = \sigma_x \otimes \sigma_y \otimes \sigma_z

> where :math:`\otimes` denotes the tensor product, and we can more simply denote by

Reviewer 2: sometimes called the [Kronecker product](https://en.wikipedia.org/wiki/Kronecker_product)

.. math::

    \mathcal{\hat{P}} = XYZ

> Other valid Pauli strings include ``III``, ``IXYZ``, ``IZYX``, etc. In general, a Pauli string of length ``N`` is a tensor product of ``N``

Reviewer 2: You can remove "In general" and just be more authoritative.  I also don't think you need a list of examples of valid pauli strings, I'm not sure it is instructive.

> Pauli matrices. A ``N``-length Pauli String in dense form is a :math:`2^N \times 2^N` matrix, so ``XYZ`` is a :math:`8 \times 8` matrix.

Reviewer 2: I don't think you need to mention dense form, the idea of dense versus sparse representation is more of how we write it in memory and not an actual mathematical term.

    P1.dim
    P1.n_qubits

    # Multiply two Pauli strings.
    phase, new_string = P1 @ P2

We can also do more complicated things, like compute the action of a Pauli string :math:`\mathcal{\hat{P}}` on a vector :math:`| \psi \rangle`, :math:`\mathcal{\hat{P}}| \psi \rangle`, or
compute the expectation value of a Pauli string with a state :math:`\langle \psi | \mathcal{\hat{P}} | \psi \rangle`. As a side note, in this guide we will use state and vector interchangeably:

> .. code-block:: python

Reviewer: It won't let me make suggestions for the whole block, so here's an edited version:
```python
import numpy as np
# Apply P to a state
P = fp.PauliString('XY')
state = np.array([1, 0, 0, 1], dtype=complex)
new_state = P.apply(state)

# Compute the expected value of P with respect to a state or a batch of states
value = P.expectation_value(state)

states = np.random.randn(4, 8) + 1j * np.random.randn(4, 8)
values = P.expectation_value(states)
```

>     # Returns "XYZ"

Reviewer 3: I think similar `# Returns` comment needed to describe the output of `P_np = P.to_tensor()`. because it's easy to confuse it with `# Returns "XYZ"`

    # Returns "XYZ"
    P_str = str(P)

For more details on the ``PauliString`` class, see the :doc:`python_api` or :doc:`cpp_api` documentation.

Pauli Operators
------------------------

The ``PauliOp`` class lets us represent operators that are linear combinations of Pauli strings with complex coefficients.
In physics, an operator is represented by a matrix in a given basis.

> For example, in the Pauli basis, we can represent an arbitrary operator :math:`A` as a sum of Pauli strings :math:`P_i` with complex coefficients :math:`c_i`:

Reviewer 2: Oo I'm not sure about this "pauli basis".  I would just say "we can represent any arbitrary operator" and not mention some "pauli basis"

> For example, in the Pauli basis, we can represent an arbitrary operator :math:`A` as a sum of Pauli strings :math:`P_i` with complex coefficients :math:`c_i`:

Reviewer 2: In your examples I would make it clear that the coeff array is c_i and the pauli string array is P_i to link it back with the math.

    A.dim
    A.n_pauli_strings

Just like with ``PauliString`` objects, we can apply ``PauliOp`` objects to a set of vectors, or compute expectation values, as well as arithmetic

> operations and dense matrix conversions. Just like with ``PauliString`` objects, we can also convert ``PauliOp`` objects back to dense numpy arrays if we'd like

Reviewer 3: `... and dense matrix conversions` - this is discussed twice, in the first and in the second sentences

    s = fp.PauliString('XYZ')
    A4 = A1 + s

>     # Apply A to a state / vector or set of states

Reviewer: ```suggestion
    # Apply A to a single state / vector or set
```

>     states = np.random.rand(10, 8) + 1j * np.random.rand(10, 8)

Reviewer: ```suggestion
    states = np.random.rand(8, 10) + 1j * np.random.rand(8, 10)
```

    P = fp.PauliString('XYZ')

>     qiskit_pauli = fp.to_qiskit(P)

Reviewer 3: in Qiskit Integration section we should replace `Fast-Pauli` entries with  `fast_pauli` to be more consistent

> For more details on Qiskit conversions, see the :doc:`python_api` or :doc:`cpp_api` documentation.

Reviewer 3: I think it's okay to just reference python api here
